stages/files: sort entries by resolved depth

ajeddeloh opened the pull request:
When sorting entries for creation order, first resolve any symlinks to
determine the actual depth.

Needs tests, draft for now

arithx commented on internal/exec/stages/files/filesystemEntries.go:
```diff
@@ -174,8 +177,8 @@ func (tmp linkEntry) create(l *log.Logger, u util.Util) error {
-func (s stage) getOrderedCreationList(config types.Config) []filesystemEntry {
-	entries := []filesystemEntry{}
+func (s stage) getOrderedCreationList(config types.Config) (entries []filesystemEntry, err error) {
+	entries = []filesystemEntry{}
```
This seems unnecessary.

ajeddeloh replied:
It is if we want the sorting function to be able to set the return value of this function

arithx replied:
Interesting, didn't realize that the sorting function couldn't refer to variables defined in the function returns.

arithx commented on internal/exec/stages/files/filesystemEntries.go:
```diff
@@ -186,12 +189,24 @@ func (s stage) getOrderedCreationList(config types.Config) []filesystemEntry {
+	sort.Slice(entries, func(i, j int) bool {
+		ipath, err2 := s.JoinPath(entries[i].getPath())
+		if err2 != nil && err == nil {
+			err = err2
```
Is there a reason that both of these cases don't immediately return?

ajeddeloh replied:
they're in a closure.

arithx replied:
Yes, but in the failure case it doesn't seem like it matters if you return an immediate `true` or `false` and it will also prevent `util.Depth` from receiving a blank string (which it doesn't seem to like as it doesn't handle the case where `path.Clean` returns `.` when it receives an empty string).

ajeddeloh replied:
Returning early isn't super useful here. That function will be called over and over. It'd be bogus results either way since either we return a static true/false that messes with ordering or we return the bogus result of comparing something via bad calls to util.Depth.

We could panic and recover but that's also not great.

We could resolve everything now changing the paths in the dir/link/file structs and not when running, but that's a much bigger refactor.

ajeddeloh commented:
Pushed a rework of this. Added tests too. Moved the generation of paths relative to the sysroot to the start of the files stage rather than right before creation.

jlebon commented on internal/exec/util/unit.go:
```diff
@@ -100,13 +107,22 @@ func (u Util) EnableRuntimeUnit(unit types.Unit, target string) error {
+	if err != nil {
+		return err
+	}
+	targetPath, err2 := u.JoinPath("/", SystemdRuntimeUnitsPath(), unit.Name)
```
Minor/optional: we're using `:=` here anyway, so might as well make it `err` here too, right?

ajeddeloh replied:
Got my variable shadowing rules mixed up. :+1: to this

ajeddeloh commented on internal/exec/util/file.go:
```diff
@@ -105,13 +105,10 @@ func (u Util) PrepareFetch(l *log.Logger, f types.File) *FetchOp {
 
 	if err := MkdirForFile(path); err != nil {
-		return err
+		return fmt.Errorf("Could not create leading directories: %v", err)
```
Wrongs `errors` package. That's from https://github.com/pkg/errors

arithx commented on internal/exec/util/unit.go:
```diff
@@ -32,7 +31,7 @@ const (
 )
 
-func FileFromSystemdUnit(unit types.Unit, runtime bool) (*FetchOp, error) {
+func (ut Util) FileFromSystemdUnit(unit types.Unit, runtime bool) (*FetchOp, error) {
```
Minor: should we either use `ut Util` in all functions and or `u Util` for consistency?
